File: solutions/python/matching-brackets/1/matching_brackets.py
def is_paired(input_string):
    opening = ('(', '{', '[')
    closing = (')', '}', ']')
    stack = []
    for input in input_string:
        if input in closing:
            if not stack or closing.index(input) != stack.pop():
                return False
        if input in opening:
            stack.append(opening.index(input))
    return not stack

File: solutions/c/matching-brackets/1/matching_brackets.c
#include "matching_brackets.h"

#include <string.h>

static int translate(const char ch) {
    switch (ch) {
        case '(':
            return 1;
        case ')':
            return -1;
        case '{':
            return 2;
        case '}':
            return -2;
        case '[':
            return 3;
        case ']':
            return -3;
        default:
            return 0;    
    }
}

bool is_paired(const char *input) {
    int stack[strlen(input)];
    size_t stack_idx = 0;
    for (char ch = *input; ch; ch = *(++input)) {
        int i = translate(ch);
        if (i > 0) {
            stack[stack_idx++] = i;
        } else if (i < 0) {
            if (i + stack[--stack_idx] != 0) {
                return false;
            }
        }
    }
    return stack_idx == 0;
}